proteuslib/unit_models/tests/test_reverse_osmosis_1D.py

```diff
@@ -293,4 +565,843 @@ def test_solution(self, RO_frame):
+            obj_str = obj.local_name
+            if obj_str[0] == '_':
+                continue  # do not test hidden references
+            assert obj_str in unit_objs_type_dict
+
+        # test feed-side control volume and associated stateblocks
+        assert isinstance(m.fs.unit.feed_side, ControlVolume1DBlock)
+        cv_stateblock_lst = ['properties']
+        for sb_str in cv_stateblock_lst:
+            sb = getattr(m.fs.unit.feed_side, sb_str)
+            assert isinstance(sb, props.NaClStateBlock)
+
+        stateblock_lst = ['permeate_side', 'permeate_out']
+        for sb_str in stateblock_lst:
+            sb = getattr(m.fs.unit, sb_str)
+            assert isinstance(sb, StateBlock)
+            assert isinstance(sb, props.NaClStateBlock)
+
+        # test statistics
+        assert number_variables(m) == 1129
+        assert number_total_constraints(m) == 1080
+        unused_list = unused_variables_set(m)
+        [print(i) for i in unused_list]
+        assert number_unused_variables(m) == 25
+
+        assert_units_consistent(m.fs.unit)
+
+        assert degrees_of_freedom(m) == 0
+
+        m.fs.properties.set_default_scaling('flow_mass_phase_comp', 1, index=('Liq', 'H2O'))
+        m.fs.properties.set_default_scaling('flow_mass_phase_comp', 1e2, index=('Liq', 'NaCl'))
+
+        set_scaling_factor(m.fs.unit.feed_side.properties[0, 1].flow_mass_phase_comp['Liq', 'NaCl'], 1e4)
+        calculate_scaling_factors(m)
+
+        # check that all variables have scaling factors
+        unscaled_var_list = list(unscaled_variables_generator(m))
+        assert len(unscaled_var_list) == 0
+        # check that all constraints have been scaled
+        unscaled_constraint_list = list(unscaled_constraints_generator(m))
+        assert len(unscaled_constraint_list) == 0
+
+        initialization_tester(m)
+
+        badly_scaled_var_lst = list(badly_scaled_var_generator(m))
+        assert badly_scaled_var_lst == []
+
+        solver.options = {'nlp_scaling_method': 'user-scaling'}
+        results = solver.solve(m)
+
+        # Check for optimal solution
+        assert_optimal_termination(results)
+
+        # Check mass conservation
+        b = m.fs.unit
+        comp_lst = ['NaCl', 'H2O']
+
+        flow_mass_inlet = sum(
+            b.feed_side.properties[0, 0].flow_mass_phase_comp['Liq', j] for j in comp_lst)
+        flow_mass_retentate = sum(
+            b.feed_side.properties[0, 1].flow_mass_phase_comp['Liq', j] for j in comp_lst)
+        flow_mass_permeate = sum(
+            b.permeate_out[0].flow_mass_phase_comp['Liq', j] for j in comp_lst)
+
+        assert value(flow_mass_inlet) == pytest.approx(1.0, rel=1e-3)
+        assert value(flow_mass_retentate) == pytest.approx(0.6103, rel=1e-3)
+        assert value(flow_mass_permeate) == pytest.approx(0.3898, rel=1e-3)
+
+        assert (abs(value(flow_mass_inlet - flow_mass_retentate - flow_mass_permeate))
+                <= 1e-2)
+
+        # Test solution
```

Comment:
We should also test the total pressure drop for the cases. And the inlet flux.

Reply:
Will do. Just a note- the "inlet" flux (first flux computed) is located one index after `length_domain.first()` since the constraint is skipped at the actual inlet. Does that make sense?